controllers/crds/cninode_controller.go

```diff
@@ -156,6 +154,17 @@ func (r *CNINodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 				}
 				shouldPatch = true
 			}
+
+			// add node id tag if it does not exist
+			if nodeId, ok := cniNode.Spec.Tags[config.NetworkInterfaceNodeIDKey]; !ok || nodeId == "" {
+				nodeId, err := r.GetNodeID(ctx, cniNode.Name)
+				if err != nil {
+					r.log.Error(err, "failed to get node id for CNINode, will retry", "cniNode", cniNode.Name)
+				} else {
+					cniNodeCopy.Spec.Tags[config.NetworkInterfaceNodeIDKey] = nodeId
+					shouldPatch = true
```

Comment:
Can we maintain a deepcopy of cninode `Spec` before doing all these and compare "possibly updated" cninode with the original copy to decide if we need to update? I would recommend to get rid of all these `shouldPatch` things...